composer/datasets/in_context_learning_evaluation.py

```diff
@@ -103,117 +108,171 @@ def _get_fewshot_sample_idxs(dataset_size: int, num_fewshot: int, sample_idx: in
-        prompt_and_fewshot = prompt_string
+        few_shot_text = preamble
 
         if num_fewshot > 0:
-            fewshot_idxs = _get_fewshot_sample_idxs(len(self.samples), num_fewshot, sample_idx, fewshot_rng)
+            fewshot_idxs = _get_fewshot_sample_idxs(len(self.dataset), num_fewshot, sample_idx, fewshot_rng)
             for fewshot_idx in fewshot_idxs:
-                context = self.samples[fewshot_idx]['context']
-                chain_of_thought = self.samples[fewshot_idx].get('chain_of_thought', '')
-                answer = self.samples[fewshot_idx]['answer']
-
-                if len(chain_of_thought) == 0:
-                    cot_delimiter = ''
-                context = f'{question_prelimiter}{context}'
-                if len(prompt_and_fewshot) > 0:
-                    context = f'{example_delimiter}{context}'
-                prompt_and_fewshot += f'{context}{continuation_delimiter}{chain_of_thought}{cot_delimiter}{answer}'
-
-        return prompt_and_fewshot
-
-    def _prep_examples(self,
-                       num_fewshot: int,
-                       prompt_string: str,
-                       example_delimiter: str,
-                       continuation_delimiter: str,
-                       question_prelimiter: str,
-                       fewshot_rng: random.Random,
-                       cot_delimiter: str = '') -> List[Dict[str, Any]]:
+                ctxt = self.construct_context(self.dataset[fewshot_idx], few_shot_text, add_answer=True)
+                few_shot_text += ctxt
+
+        return few_shot_text
+
+    def construct_context(self, sample: dict, preceding_text: str = '', add_answer: bool = False):
+        ctxt = sample[self.context_key]
+        ctxt = f'{self.prelimiter}{ctxt}'
+        if len(preceding_text) > 0:
+            ctxt = f'{self.example_delimiter}{ctxt}'
+        ctxt = f'{ctxt}{self.continuation_delimiter}'
+        if add_answer:
+            ctxt = f'{ctxt}{self.get_answer_from_sample(sample)}'
+        return ctxt
+
+    def get_answer_from_sample(self, sample: dict):
+        return sample[self.answer_key]
+
+    def fix_eos_on_preamble(self, preamble: dict):
+        # If the preamble is empty then this will be a 0-length list, unless the tokenizer adds special tokens to empty strings (e.g. OPT tokenizer)
+        # If there is an EOS token added, we need to remove it so it is not in the middle of the prompt
+        if (self.tokenizer.eos_token_id is not None and len(preamble['input_ids']) > 1 and
+                preamble['input_ids'][-1] == self.tokenizer.eos_token_id):
+            preamble['input_ids'] = preamble['input_ids'][:-1]
+        return preamble
+
+    def tokenize_example(self, prompt_and_fewshot: str, ctxt: str, example: dict):
+        tokenized_example = {}
+        preamble = self.tokenizer(prompt_and_fewshot)
+        preamble = self.fix_eos_on_preamble(preamble)
+        tokenized_example['preamble'] = preamble
+        if self.strip_data:
```

Comment:
why does this need to be conditional at all?

Reply:
we don't want to remove whitespace from code data